---
title: 异步及异步编程的解决方案
date: 2020-04-07 10:48:32
tags: JavaScript
categories: JavaScript
---

## 什么是异步
简单的来说就是一件事不是连续紧接着做完的，可以分为几个阶段，**先进行一个阶段，剩下的阶段延后执行**。在延迟的这段时间，做的是别的任务。相应的，连续执行的任务是同步的

例如：
```
setTimeout(function() {
    console.log("4")
}, 0);
console.log("5");
```
`setTimeout`就是一个异步任务，先做延迟，然后在延迟期间把执行权交给了`console.log`，所以是先输出5，再输出4.

为什么要异步呢，因为JavaScript是单线程语言，如果像上面那种情况，还是使用同步的话，那延迟期间，线程里什么事情都没有做，会白白浪费处理器时间，耗费性能。

## 异步编程方案
异步编程有很多解决的方案，其演变过程是： 回调函数=>`promise`=> `Generator` => `async/await`,每个演变过程都会解决一些之前的问题

### 回调函数

举个栗子，Node里的fs模块，读取文件的操作：
```
fs.readFile('url','utf-8',function(err,data){
    if(err) throw err;
    console.log(data);
})
```
上面的代码中，readFile函数的第三个参数，就是回调函数，也就是等操作系统返回对应的文件之后，才会执行这个回调函数的操作。

也即是说，在执行`readFile`的时候，执行权交给了主线程执行栈的同步任务，同步任务执行完之后，才会执行回调函数对应的操作。

#### 回调函数的缺点
回调函数会导致回调地狱。

比如说现在有很多异步任务，且任务有依赖关系（一个任务需要拿到另一个任务成功后的结果才能开始执行）的时候，回调的方式写出来的代码就会像这样：
```
getData1(data1 => {
  getData2(data1, data2 => {
    getData3(data2, data3 => {
      getData4(data3, data4 => {
        getData5(data4, data5 => {
          // 终于取到data5了
        })
      })
    })
  })
})
```
这种多层嵌套的结构就是回调地狱，**代码可读性很差**
## Promise
promise的出现就是为了解决回调函数回调地狱的问题。**将回调函数的嵌套改为了链式调用**

Promise可以当作一个状态机，promise实例有三个状态：
* Pending(进行中)
* Fulfilled(已成功)
* rejected (已失败)

我们使用`new Promise`去创建一个Promise实例，这个Promise实例会传入一个函数作为参数，函数又有两个函数作为参数，分别是：`resolve`、`reject`。

执行`resolve`函数，`Promise`实例的状态会变为`fulfilled`，后续就会去执行.then回调函数

执行`reject`函数，`Promise`实例的状态会变为`rejected`，后续就会去执行`.catch`回调函数，或者`.then`的第二个回调函数。

基本结构
```
let p = new Promise((resolve,reject)=>{
    //做一些事情
    //然后在某些条件下resolve，或者reject（以下代码）
    if(/* 条件随便写 */){
       	 resolve()
     } else{
         reject()
     }
})

p.then(()=>{
    //如果p的状态被resolve了，就进入这里
},()=>{
    //如果p的状态被reject
})//这么写reject，使用了then的第二个回调函数
```
`promise`的异步体现在，`new promise`的部分是立即执行的，执行完毕之后，会把执行权交给主线程的执行栈执行同步任务，然后再是执行，`then`和`catch`的回调。这里涉及到了JavaScript的事件执行机制

### promise 实现多任务顺序执行
```
var result=new Promise(function(resolve,reject){
	setTimeout(function(){
		resolve("one");
	},3000)
}).then(function(data){
	console.log(data);
	return new Promise(function(resolve,reject){
		setTimeout(function(){
			resolve("two");
		},3000)
	})
}).then(function(data){
	console.log(data);
})
```
原理是在第一个`promise`实例的`then`函数回调中再return一个`promise`实例，也就是后续的任务
### promise 实现多任务并行执行
```
let p = new Promise((resolve,reject)=>{
    //做一些事情
    //然后在某些条件下resolve，或者reject（以下代码）
    if(/* 条件随便写 */){
       	 resolve()
     } else{
         reject()
     }
})

p.then(()=>{
    //如果p的状态被resolve了，就进入这里
},()=>{
    //如果p的状态被reject
})
```
这里主要使用了`Promise.all`这种结构，`Promise.all([p1,p2])`接受一个数组作为参数，数组的元素都是`Promise`实例，只有当数组中所有的`promise`实例的状态都变为`fulfilled`的时候，这个`Promise.all`的实例才会变为`fulfilled`，才能执行后续的`.then`操作。

和`Promise.all`相关的还有`Promise.race`，那`Promise.race`实例的状态是等于其参数中第一个执行完毕的`Promise`实例的状态，它有可能是`rejected`，也有可能是`fulfilled`。

### promise的缺点
`promise`的最大的缺点就是会造成代码冗余，原来的任务被`promise`包装了一下，不管什么操作，一眼看过去都是一堆的then,原来的语义变得很不清楚。通俗点讲，就是链式调用得多了，then多了，你都不知道第一个函数是干嘛的了

## Generator
Generator函数是ES6提供的异步编程解决方案，generator解决的是上面promise代码冗余的问题。它用的是类似同步的写法来执行一些异步操作

* `generator`函数的声明方式是`function *`，函数内部使用`yield`表达式，定义不同的内部状态
* 执行Generator函数会返回一个**遍历器对象**
* 调用Generator函数后，该函数并不执行，返回不是函数运行结果，而是一个执行内部状态的指针
* 必须调用遍历器对象的next方法，使得指针移向下个状态，输出返回的结果。

### Generator函数的写法

```
function* helloWorldGenerator() {
  yield 'hello';
  yield 'world';
  return 'ending';
}

var hw = helloWorldGenerator();
hw.next()
// { value: 'hello', done: false }

hw.next()
// { value: 'world', done: false }

hw.next()
// { value: 'ending', done: true }

hw.next()
// { value: undefined, done: true }
```
### 异步应用
因为`yield`能够中断执行代码的特性，可以帮助我们来**控制异步代码的执行**

例如有两个异步函数`A`和`B`，并且B的参数是A的返回值，也就是说，如果`A`没有执行结束，我们执行结束，我们不能执行`B`

那这时候我们写一段伪代码：
```
function* effect() {
  const { param } = yield A();
  const { result } = yield B(param);
  console.table(result);
}
```
这时候我们如果需要得到result，那我们就需要：
```
const iterator = effect()
iterator.next()
iterator.next()
```
co库可以用来每次执行`A()/b()`请求结束之后，都会自动执行`next()`方法

### 使用Generator去实现promise的任务顺序执行

Generator版本的顺序任务执行
```
function f1() {
  setTimeout(function() {
    g.next("one");  //将参数传给data1
  }, 3000);
}

function f2(data1) {
  console.log("接收到了" + data1);
  setTimeout(function(){
      g.next("two");  //将参数传给data2
  },3000)
}

function* mygenerator() {
  var data1 = yield f1();
  var data2=yield f2(data1);
  console.log(data2);
}

var g = mygenerator();
g.next();
```
这里要注意的一个点是：执行`yield f1()`的时候是不会返回的结果返回给`data1`的。next方法可以接受参数，这是向`Generator`函数体内输入数据。第二个`next`的时候传入参数，就能被变量`data1`就能够接收到。  
而且不能在f1中直接调用`g.next()`，要在异步函数内调用
再看一个栗子理解:
```
function* G() {
    const a = yield 100
    console.log('a', a)  // a aaa
    const b = yield 200
    console.log('b', b)  // b bbb
    const c = yield 300
    console.log('c', c)  // c ccc
}
const g = G()
g.next()    // value: 100, done: false
g.next('aaa') // a aaa   value: 200, done: false   
g.next('bbb') //   b bbb  value: 300, done: false
g.next('ccc') //c ccc  value: undefined, done: true
```
### 捕获错误
```
function* gen(x){
  try {
    var y = yield x + 2;
  } catch (e){
    console.log(e);
  }
  return y;
}

var g = gen(1);
g.next();
g.throw('出错了');
// 出错了
```
使用`try...catch`。

## async和await
async函数就是Generator的语法糖。  
形式上的不同：
* `async`函数将`Generator`函数的星号(*)替换成`async`
* 将`yield`替换成`await`


async函数对Generator函数改进体现在以下三点： 


1.内置执行器

也就是说async函数的执行，和普通函数一样，只需要一行就可以。不用像Generator函数需要调用`next`方法才能真正执行。

例如对于一个async函数来说:
```
const asyncReadFile = async function () {
  const f1 = await readFile('/etc/fstab');
  const f2 = await readFile('/etc/shells');
  console.log(f1.toString());
  console.log(f2.toString());
};
```
调用时只需要：
```
asyncReadFile()
```
2.更好的语义
`async`和`await`比起星号和`yield`，语义更加清楚了。`async`表示函数里有异步操作，`await`表示紧跟在后面的表达式需要等待结果。  
3.返回值是Promise  
`async`函数的返回值是 `Promise` 对象，这比 `Generator` 函数的返回值是 `Iterator `对象方便多了。你可以用then方法指定下一步的操作。
### async函数的基本用法
`async`函数返回一个` Promise `对象，可以使用`then`方法添加回调函数。当函数执行的时候，一旦遇到`await`就会先返回，等到异步操作完成，再接着执行函数体内后面的语句。
```
async function getStockPriceByName(name) {
  const symbol = await getStockSymbol(name);
  const stockPrice = await getStockPrice(symbol);
  return stockPrice;
}

getStockPriceByName('goog').then(function (result) {
  console.log(result);
});
```
先执行了第一个`await`后的`getStockSymbol(name)`函数；得到了股票的名称`symbol`后，将`symbol`传给第二个`await`后面的`getStockPrice(symbol)`作为参数；最后返回股票价格`stockPrice`。

### async/await错误处理
`try/catch`
使用`try`将`await`语句包含起来，如果`await`后的语句执行错误，则错误会被`catch`捕获：
```
run();

async function run() {
    try {
        await Promise.reject(new Error("Oops!"));
    } catch (error) {
        error.message; // "Oops!"
    }
}
//Oops!
```
### 执行顺序问题
```
async function async1(){
   console.log('async1 start');
    await async2();
    console.log('async1 end')
}
async function async2(){
    console.log('async2')
}

console.log('script start');
async1();
console.log('script end')

// 输出顺序：script start->async1 start->async2->script end->async1 end
```
当`async`函数执行的时候，一旦遇到`await`就会先返回，等到触发的异步操作完成后，再执行函数体后面的语句。可以理解为，是让出了线程，跳出了`async`函数体。
```
async function async1() {
    console.log('async1 start');
    await async2();
    console.log('async1 end');
}
async function async2() {
    console.log('async2');
}
console.log('script start');
setTimeout(function() {
    console.log('setTimeout');
}, 0)
async1();
new Promise(function(resolve) {
    console.log('promise1');
    resolve();
}).then(function() {
    console.log('promise2');
});
console.log('script end');
```
执行顺序为：
```
script start-async1 start-async2-promise1-script end-async1 end-promise2--setTimeout
```
执行`async1`的时候遇到`await`先去执行`async2`，然后跳出了函数体，去执行后续的代码，然后再回到当前`async1`函数当中执行`await`后续语句。

## 参考
https://delaprada.com/2020/01/14/Generator%E5%87%BD%E6%95%B0-async%E5%87%BD%E6%95%B0/#more
